src/KubeOps.Operator/Builder/OperatorBuilder.cs

```diff
@@ -59,162 +58,36 @@ public OperatorBuilder(IServiceCollection services, OperatorSettings settings)
-                                 ReportingComponent = settings.Name,
-                                 ReportingInstance = Environment.MachineName,
-                                 Source = new() { Component = settings.Name, },
-                                 InvolvedObject = entity.MakeObjectReference(),
-                                 FirstTimestamp = DateTime.UtcNow,
-                                 LastTimestamp = DateTime.UtcNow,
-                                 Count = 0,
-                             }.Initialize();
-
-                @event.Count++;
-                @event.LastTimestamp = DateTime.UtcNow;
-                logger?.LogTrace(
-                    "Save event with new count {count} and last timestamp {timestamp}",
-                    @event.Count,
-                    @event.LastTimestamp);
-
-                try
-                {
-                    await client.SaveAsync(@event);
-                    logger?.LogInformation(
-                        """Created or updated event with name "{name}" to new count {count} on entity "{kind}/{name}".""",
-                        eventName,
-                        @event.Count,
-                        entity.Kind,
-                        entity.Name());
-                }
-                catch (Exception e)
-                {
-                    logger?.LogError(
-                        e,
-                        """Could not publish event with name "{name}" on entity "{kind}/{name}".""",
-                        eventName,
-                        entity.Kind,
-                        entity.Name());
-                }
-            };
-
     private void AddOperatorBase()
     {
         Services.AddSingleton(_settings);
-        Services.AddTransient<IKubernetesClient>(_ => new KubernetesClient.KubernetesClient());
-        Services.AddTransient(CreateEventPublisher());
+
+        // Add the default configuration and the client separately. This allows external users to override either
+        // just the config (e.g. for integration tests) or to replace the whole client, e.g. with a mock.
+        Services.TryAddSingleton(KubernetesClientConfiguration.BuildDefaultConfig());
+        Services.TryAddTransient<IKubernetesClient>(services =>
+            new KubernetesClient.KubernetesClient(services.GetRequiredService<KubernetesClientConfiguration>()));
+
+        // We also add the k8s.IKubernetes as a transient service, in order to allow to access internal services
+        // and also external users to make use of it's features that might not be implemented in the adapted client.
+        Services.TryAddTransient<IKubernetes>(services =>
+            new Kubernetes(services.GetRequiredService<KubernetesClientConfiguration>()));
+        Services.TryAddTransient<IEventPublisherFactory, KubeOpsEventPublisherFactory>();
+        Services.TryAddTransient<EventPublisher>(
+            services => services.GetRequiredService<IEventPublisherFactory>().Create());

```

Comment:
I do like the addition of the client/config/original-client separatly!

The factories, I'm not so sure about. I don't see the benefit of having factories for the delegates.

Reply:
Is it to allow configuration from outside I guess... ? :)